[ATOM-16021] Initial continuous capture support

jcbhl opened the pull request:
### Please [start a build](https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-2624/) and request sig-presentation!

This PR implements capturing multiple frames of CPU profiling data and saving it to disk so that it can be analyzed offline (which will be implemented in the visualizer in a future PR). Now, the ProfilingCaptureSystemComponent offers endpoints for capturing one frame of profiling data (with a blocking IO call) in addition to an arbitrary amount of frames (using an IO thread). 


### Demo
https://user-images.githubusercontent.com/64656371/127545500-06bed756-0d15-48ff-8966-6ddf402af690.mp4
![image](https://user-images.githubusercontent.com/64656371/127545626-4b6b4e27-bce1-4370-a228-65fc5d584750.png)
Resulting file size on disk was 70 MB from ~3000 frames of data.

jcbhl commented on Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:
```diff
@@ -590,6 +600,50 @@ namespace AZ
+                return false;
+            }
+
+            // cpuProfilingData could be 1GB+ once saved, so use an IO thread to write it to disk.
```
Note: please be super critical of my use of threading here, I haven't used threading primitives before so my use is likely incorrect 😁

jcbhl commented on Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.h:
```diff
@@ -84,6 +87,13 @@ namespace AZ
+
+            // Stores the last continuous capture data so that the IO thread can access it by reference. 
+            // Ensure that this is not cleared until m_cpuDataSerializationInProgress is false. 
+            AZStd::deque<RHI::CpuProfiler::TimeRegionMap> m_lastContinuousCapture;
```
Note: I looked into the lifetime of `ProfilingCaptureSystemComponent` and it appears that the component is only destructed when the renderer is destructed (see `CommonModule.cpp`) so this deque should always be alive while the IO thread is working.

jcbhl commented on Gems/Atom/RHI/Code/Include/Atom/RHI/CpuProfilerImpl.h:
```diff
@@ -134,6 +139,12 @@ namespace AZ
+
+            // Stores multiple frames of profiling data, size is controlled by MaxFramesToSave. Flushed when EndContinuousCapture is called.
+            // Deque so that we can have fast append of new data + removal of old profiling data.
+            AZStd::deque<TimeRegionMap> m_continuousCaptureData;
```
Question: thoughts about using deque here over a vector? I chose it so that we could have O(1) removal of old profiling data if we are at `MaxFramesToSave` but it may perform worse than just a plain old vector due to cache effects.

jeremyong-az replied:
You could use a circular or ring buffer to get the benefits of both O(1) removal/insertion in a linear range of memory like a vector.

jcbhl replied:
That's exactly what I was looking for, thanks! Moved to `AZStd::ring_buffer`

jeremyong-az commented on Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:
```diff
@@ -590,6 +600,50 @@ namespace AZ
+                    flag.store(false);
+                });
+
+            thread.detach(); 
```
It looks like you're joining this thread later, so you shouldn't detach the thread at this point I don't think

jeremyong-az commented on Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:
```diff
@@ -590,6 +600,50 @@ namespace AZ
+
+            // cpuProfilingData could be 1GB+ once saved, so use an IO thread to write it to disk.
+            m_cpuDataSerializationInProgress.store(true);
+            auto thread = AZStd::thread(
```
Do we not have a thread pool for enqueuing async work of this nature as opposed to spinning up a dedicated thread, or is the idea to keep work off the async thread pool for some reason?

jcbhl replied:
No specific reason for using an owned thread here - moved to `AZ::Job` which should avoid some of the overhead of spawning a new thread!

jeremyong-az commented on Gems/Atom/RHI/Code/Source/RHI/CpuProfilerImpl.cpp:
```diff
@@ -141,12 +144,47 @@ namespace AZ
 
+        bool CpuProfilerImpl::BeginContinuousCapture()
+        {
+            if (m_continuousCaptureInProgress)
```
After loading the value in this atomic, nothing stops a different thread from entering the same place and setting `m_continuousCaptureInProgress` to true on line 156, before _this thread_ gets to the same line. While this is technically benign because the atomic will stay true in both cases, you might have two threads competing to deactivate the capture later because they both believe they started the capture.

The pattern to employ here is the atomic exchange, where you do something like

```cpp
if (m_continuousCaptureInProgress.compare_exchange_strong(false, true))
{
    // We atomically transitioned the value of m_continuousCaptureInProgress from false to true, so are responsible for the capture and can start it
}
else
{
    return false;
}
```

jcbhl commented:
Thanks again for pointing out the issues here @jeremyong-az. I went back on unifying the state of the CpuProfiler and the ProfilingCaptureSystemComponent - I want to separate the end use of the data from the internal implementation, as some devs might want to do in-memory work with the data that doesn't require writing to disk (ex. calculating aggregate statistics). 


I switched to using `compare_exchange_strong` for the atomics like you suggested, and also added a mutex around `EndContinuousCapture()`. I realized that multiple threads could enter that function and move out of `m_continuousCaptureData` more than once, which I believe is UB. Also added [[nodiscard]] to better indicate that the functions may fail, ideally this would be an `AZ::Outcome` but I wanted to keep the changes small here.

jeremyong-az commented on Gems/Atom/Feature/Common/Code/Source/ProfilingCaptureSystemComponent.cpp:
```diff
@@ -603,55 +603,49 @@ namespace AZ
-                m_cpuDataSerializationThread.join();
-            }
+                auto thread = AZStd::thread(threadIoFunction);
+                m_cpuDataSerializationThread = AZStd::move(thread);
```
Random nit, you can just do `m_cpuDataSerializationThread = AZStd::thread{threadIoFunction}` here.

jeremyong-az reviewed (approved):
I will say, I'm not 100% convinced myself that this is necessarily thread safe from just a visual code inspection myself, as there are still a number of complex interactions. I will approve but recommend independent testing or simplification in cases where the atomic + mutex interaction isn't strictly necessary for performance.

jcbhl commented:
Definitely agree - this is pretty messy at the moment. I'll put up a jira for adding automated tests + cleaning this up. Thanks!
